catalyst/contrib/data/cv/datasets/image.py

```diff
@@ -0,0 +1,324 @@
+            )
+            self.margins.append(margins)
+            image_shape_ext = Shape(
+                height=shape.height + margins.bottom + margins.top,
+                width=shape.width + margins.left + margins.right,
+            )
+
+            x_stop = image_shape_ext.width - overlap.width
+            y_stop = image_shape_ext.height - overlap.height
+
+            x = torch.arange(0, x_stop, self.tile_step.width)
+            y = torch.arange(0, y_stop, self.tile_step.height)
+
+            crops = [
+                tuple(map(lambda t: t.item(), pair))
+                for pair in torch.cartesian_prod(x, y)
+            ]
+            self.crops.append(crops)
+
+    def _init_index(self):
+        self.num_tiles_cumulative = []
+        cumulative = 0
+        for crops in self.crops:
+            num_tiles = len(crops)
+            cumulative += num_tiles
+            self.num_tiles_cumulative.append(cumulative)
+
+    def _extend_image_with_margins(self, image: np.ndarray, margins: Edges):
+        border_type = cv2.BORDER_DEFAULT if self.train else cv2.BORDER_CONSTANT
+        image = cv2.copyMakeBorder(
+            image,
+            margins.top,
+            margins.bottom,
+            margins.left,
+            margins.right,
+            borderType=border_type,
+        )
+
+        return image
+
+    def __len__(self):
+        """
+        Returns length of dataset, i.e. number of tiles.
+        """
+        return self.num_tiles_cumulative[-1]
+
+    def __getitem__(self, idx: int):
+        """
+        Args:
+            idx: index of element in dataset
+
+        Returns:
+            OrderedDict with tile and starting x, y coordinates.
+        """
+        image_idx = bisect_right(self.num_tiles_cumulative, idx)
+        crop_idx = idx - self.num_tiles_cumulative[image_idx]
+        x, y = self.crops[image_idx][crop_idx]
+        margins = self.margins[image_idx]
+
+        image = skimage.io.imread(self.images[image_idx])
+        image = self._extend_image_with_margins(image, margins)
+        height, width, *_ = image.shape
+        storage_size = Shape(height=height, width=width)
+        image = image[
+            y : y + self.tile_size.height, x : x + self.tile_size.width
+        ]
+        dict_ = {"image": image}
+
+        if self.masks is not None:
+            mask = skimage.io.imread(self.masks[image_idx])
+            mask = self._extend_image_with_margins(mask, margins)
+            mask = mask[
+                y : y + self.tile_size.height, x : x + self.tile_size.width
+            ]
+            dict_.update(mask=mask)
+
+        if self.transform is not None:
+            if isinstance(self.transform, (BasicTransform, BaseCompose)):
+                transform = self.transform(**dict_)
+            else:
+                transform = self.transform(dict_)
+            image = transform["image"]
+            if self.masks is not None:
+                mask = transform["mask"]
```

Comment:
In my opinion, it's better to assume that transform work with dicts only and return dict
and in case of albumentations/kornia/whatever user can use `Augmentor`s